lib/temporalio/worker.rb

```diff
@@ -41,38 +48,67 @@ def initialize(
       end
     end
 
-    def run
-      Async { |task| start(task) }
+    def run(&block)
+      Runner.new(self).run(&block)
     end
 
-    def start(reactor = nil)
+    def start(runner = nil)
       mutex.synchronize do
-        raise 'Worker is already running' if running?
+        raise 'Worker is already started' if started?
 
-        @running = true
+        @started = true
       end
 
+      @runner = runner
       runtime.ensure_callback_loop
-      reactor ||= runtime.reactor
-      reactor.async { |task| activity_worker.run(task) } if activity_worker
-      reactor.async { |task| workflow_worker.run(task) } if workflow_worker
-    end
 
-    def shutdown
-      core_worker.initiate_shutdown
-      activity_worker&.shutdown
-      workflow_worker&.shutdown
-      core_worker.shutdown
+      runtime.reactor.async do |task|
+        if activity_worker
+          task.async do |task|
+            activity_worker.run(task)
+          rescue StandardError => e
+            shutdown(e) # initiate shutdown because of a fatal error
+          end
+        end
+
+        # TODO: Pending implementation
+        task.async { |task| workflow_worker.run(task) } if workflow_worker
+      end
     end
 
-    private
+    def shutdown(exception = Temporalio::Error::WorkerShutdown.new('Manual shutdown'))
+      mutex.synchronize do
+        return unless running?
 
-    attr_reader :mutex, :runtime, :core_worker, :activity_worker, :workflow_worker
+        # First initiate Core shutdown, which will start dropping poll requests
+        core_worker.initiate_shutdown
+        # Then let the runner know we're shutting down, so it can stop other workers
+        runner&.shutdown(exception)
+        # Wait for workers to drain any outstanding tasks
+        activity_worker&.drain
+        workflow_worker&.drain
```

Comment:
This is just waiting for the workers to be drained, right? It doesn't actually initiate the operation, right?

Reply:
Yes, that's right!